README.md
# `monitoring-translation` component

This component is part of the 5G-MEDIA MAPE service. Take a look in the [MAPE](https://github.com/5g-media/mape) repository.

## Introduction
The `monitoring-translation` component:
- consumes raw monitoring metrics from specific topics in the Apache Kafka broker,
- retrieves details for the involved VDU through the OSM REST API (SO-ub), 
- generates a specific structure for any raw metric related with any origin,
- publishes the final structure of the metrics in Apache Kafka broker in the topic with pattern `ns.instances.trans`.

The role of this service is to homogenise the monitoring data in a uniform structure independent from the source origin and 
correlate them in ns-aware topics.

It supports:
 - [x] OpenStack (see `openstack.py`)
 - [x] OpenNebula (see `opennebula.py`) 
 - [x] Kubernetes (see `kubernetes.py`)
 - [x] VNFs


The indicative structure of the monitoring metric is:
```json
{
    "metric": {
        "timestamp": "datetime|iso8601",
        "value": "number",
        "type": "string",
        "name": "string",
        "unit": "string"
    },
    "mano": {
        "ns": {
            "nsd_id": "uuid",
            "name": "string",
            "id": "uuid",
            "nsd_name": "string"
        },
        "vim": {
            "tag": "string",
            "type": "string",
            "name": "string",
            "uuid": "uuid",
            "url": "string|url"
        },
        "vdu": {
            "status": "string|null",
            "name": "string|null",
            "ip_address": "string|ip",
            "mgmt-interface": "string|null",
            "image_id": "uuid|null",
            "flavor": {
                "vcpus": "number|null",
                "ram": "number|null",
                "disk": "number|null"
            },
            "id": "string|uuid|null"
        },
        "vnf": {
            "short_name": "string|null",
            "vnfd_name": "string",
            "id": "uuid",
            "name": "string|null",
            "vnfd_id": "uuid",
            "index": "integer"
        }
    }
}
```

## Requirements
- Python 3.5+ 
  + a set of python packages are used (see `requirements.txt`).
- The Apache Kafka bus must be accessible from the service
- The OSM REST APIs must be accessible from the service
- The Graylog must be accessible from the service.

## Configuration
Check the `settings.py` file:
 - *KAFKA_SERVER*: defines the *host/IPv4* and *port* of the Kafka bus.
 - *KAFKA_CLIENT_ID*: defines the id of the kafka application.
 - *KAFKA_API_VERSION*: defines the version of the Kafka bus. The default is `(1, 1, 0)`.
 - *KAFKA_GROUP_ID*: defines the consumer group of the Kafka application.
 - *KAFKA_MONITORING_TOPICS*: a dictionary that indicates the topic of the messages per monitoring data origin.
 - *KAFKA_TRANSLATION_TOPIC*: defines the topic in which the Translator publishes the messages. The default topic is `ns.instances.trans`. 
 - *KAFKA_EXECUTION_TOPIC*: defines the topic where the optimization messages are sent. The default topic is `ns.instances.exec`.
 - *OSM_IP*: defines the IPv4 of the OSM.
 - *OSM_ADMIN_CREDENTIALS*: defines the admin credentials of the OSM.
 - *OSM_COMPONENTS*: defines the OSM services.
 - *REDIS_HOST*: defines the host/IPv4 of the Redis.
 - *REDIS_PORT*: defines the port of the Redis.
 - *REDIS_NFVI_DB*: defines the db where the NFVI data are stored.
 - *REDIS_PASSWORD*: defines the password of the redis. By default, no password is needed, just `None`.
 - *REDIS_EXPIRATION_SECONDS*: indicates expiration policy of the redis entries in seconds. Default value is `86400 minutes` or `24 hours`.
 - *INFLUX_DATABASES*: defines the settings for the InfluxDB timeseries such as host, ip, db name and so on.
 - *GRAYLOG_HOST*: defines the host/IPv4 of the graylog server.
 - *GRAYLOG_PORT*: defines the port of the graylog server.
 - *LOGGING*: declares the logging (files, paths, backups, max length). By default, the logs with level WARNING, ERROR, CRITICAL are sent in Graylog over UDP.

## Installation/Deployment

To build the docker image, copy the bash script included in the `bash_scripts/` folder in the parent folder of the project and then, run:
```bash
   chmod +x build_docker_image.sh
   ./build_docker_image.sh
```

Given the docker image availability, there are 3 ways to deploy it:
- using the [MAPE docker-compose](https://github.com/5g-media/mape) project
- using the gitlab CI recipe as standalone container
- manual deployment as standalone container


For manual deployment in linux VM, download the code from the repo in the user home (e.g. `/home/ubuntu`) and follow the below steps:
```bash
# Or, instantiate the docker container using the custom values of the env variables
$ sudo docker run -p 8888:3333 --name monitoring_data_translator --restart always \
    -e DEBUG=0 \
    -e KAFKA_HOST="192.158.1.175" \
    -e OSM_IP="192.168.1.196" \
    -e OSM_USER="admin" \
    -e OSM_PWD="osmpassword" \
    -e REDIS_PORT="6379" \
    -e REDIS_EXPIRATION_SEC="14400" \
    -e KAFKA_PORT="9092" \
    -e KAFKA_TRANSLATION_TOPIC="ns.instances.trans" \
    -e INFLUXDB_DB_NAME="monitoring" \
    -e INFLUXDB_USER="root" \
    -e INFLUXDB_PWD="influxdbuser" \
    -e INFLUXDB_PORT="8086" \
    -e GRAYLOG_HOST="192.168.1.175" \
    -e GRAYLOG_PORT="12201" \    
    -dit monitoring-data-translator
```

This service is part of the docker-compose MAPE project.

## Usage

Access the docker container:
```bash
$ sudo docker exec -it  monitoring_data_translator bash
```

Start the translator service through the supervisor:
```bash
$ service supervisor start && supervisorctl start <plugin_name>
```

Stop the translator service through the supervisor:
```bash
$ supervisorctl stop <plugin_name>
```

The `plugin_name` could be:
- openstack-plugin
- kubernetes-plugin
- vcache-plugin

Stop the supervisor service:
```bash
$ service supervisor stop 
```

You are able to check the status of the plugins using your browser from the supervisor UI.
Type the URL: `http://{ipv4}:8888`


## Authors
- <pathanasoulis@ep.singularlogic.eu>

## Contributors
 - Contact with Authors
 
## Acknowledgements
This project has received funding from the European Union’s Horizon 2020 research and innovation programme under grant agreement No 761699. The dissemination of results herein reflects only the author’s view and the European Commission is not responsible for any use that may be made of the information it contains.

## License
[Apache 2.0](LICENSE.md)




CHANGELOG.md
2019-10-03
--------------
- Integration with Graylog stack for logging over UDP
- Add documentation using sphinx

2019-mm-dd
--------------
- Support k8s, one,, ... apps

2019-mm-dd
--------------
- Changes to be compatible with the OSM-r5 NBI API

2018-11-12
--------------
- Changes to be compatible with the OSM-r4 NBI API

2018-07-18
--------------
- Subscribe in pattern-oriented topic in case of OpenStack. In this way, the `openstack_runner` service can listen messages from topics such as `nvfi.eng001.openstack` and `nvfi.ote123.openstack` in parallel 

2018-05-12
--------------
- Use supervisor to manage the service
- Bug fixes

2018-05-11
--------------
- Define a standard data model for teh monitoring metrics

2018-05-10
--------------
- Initial commit

settings.py
import os

DEBUG = int(os.environ.get("DEBUG", 0))
PROJECT_ROOT = os.path.dirname(os.path.realpath(__file__))
GRAYLOG_LOGGING = int(os.environ.get("GRAYLOG_LOGGING", 0))

# =================================
# WHITE LIST OF MONITORING METRICS
# =================================
# For ceilometer, see: https://docs.openstack.org/ceilometer/latest/admin/telemetry-measurements.html
METRICS_WHITE_LIST = {
    "openstack": ['memory', 'memory.usage', 'memory.resident', 'memory.bandwidth.total',
                  'memory.bandwidth.local',
                  'memory.swap.in', 'memory.swap.out', 'memory_util',
                  'cpu', 'cpu_util', 'cpu.delta', 'vcpus', 'cpu_l3_cache', 'perf.cpu.cycles',
                  'network.incoming.bytes', 'network.incoming.bytes.rate', 'network.outgoing.bytes',
                  'network.outgoing.bytes.rate', 'network.incoming.packets',
                  'network.incoming.packets.rate',
                  'network.outgoing.packets', 'network.outgoing.packets.rate',
                  'network.incoming.packets.drop',
                  'network.outgoing.packets.drop', 'network.incoming.packets.error',
                  'network.outgoing.packets.error',
                  'disk.read.requests', 'disk.read.requests.rate', 'disk.write.requests',
                  'disk.write.requests.rate',
                  'disk.read.bytes', 'disk.read.bytes.rate', 'disk.write.bytes',
                  'disk.write.bytes.rate',
                  'disk.device.read.requests', 'disk.device.read.requests.rate',
                  'disk.device.write.requests',
                  'disk.device.write.requests.rate', 'disk.device.read.bytes',
                  'disk.device.read.bytes.rate',
                  'disk.device.write.bytes', 'disk.device.write.bytes.rate', 'disk.root.size',
                  'disk.ephemeral.size',
                  'disk.latency', 'disk.iops', 'disk.device.latency', 'disk.device.iops',
                  'disk.capacity', 'disk.usage',
                  'disk.allocation', 'disk.device.capacity', 'disk.device.allocation',
                  'disk.device.usage',
                  'disk.device.read.latency', 'disk.device.write.latency'],
    "opennebula": ['', ],
    "unikernels": ['*', ],
    "kubernetes": ['', ],
    "telegraf": ["bytes_recv", "bytes_sent", "drop_in", "drop_out", "err_in", "err_out",
                 "packets_recv", "packets_sent",
                 "icmp_inaddrmaskreps", "icmp_inaddrmasks", "icmp_incsumerrors",
                 "icmp_indestunreachs",
                 "icmp_inechoreps", "icmp_inechos", "icmp_inerrors", "icmp_inmsgs",
                 "icmp_inparmprobs",
                 "icmp_inredirects", "icmp_insrcquenchs", "icmp_intimeexcds",
                 "icmp_intimestampreps",
                 "icmp_intimestamps", "icmp_outaddrmaskreps", "icmp_outaddrmasks",
                 "icmp_outdestunreachs",
                 "icmp_outechoreps", "icmp_outechos", "icmp_outerrors", "icmp_outmsgs",
                 "icmp_outparmprobs",
                 "icmp_outredirects", "icmp_outsrcquenchs", "icmp_outtimeexcds",
                 "icmp_outtimestampreps",
                 "icmp_outtimestamps", "icmpmsg_intype3", "icmpmsg_outtype3", "ip_defaultttl",
                 "ip_forwarding",
                 "ip_forwdatagrams", "ip_fragcreates", "ip_fragfails", "ip_fragoks",
                 "ip_inaddrerrors", "ip_indelivers",
                 "ip_indiscards", "ip_inhdrerrors", "ip_inreceives", "ip_inunknownprotos",
                 "ip_outdiscards",
                 "ip_outnoroutes", "ip_outrequests", "ip_reasmfails", "ip_reasmoks",
                 "ip_reasmreqds", "ip_reasmtimeout",
                 "tcp_activeopens", "tcp_attemptfails", "tcp_currestab", "tcp_estabresets",
                 "tcp_incsumerrors",
                 "tcp_inerrs", "tcp_insegs", "tcp_maxconn", "tcp_outrsts", "tcp_outsegs",
                 "tcp_passiveopens",
                 "tcp_retranssegs", "tcp_rtoalgorithm", "tcp_rtomax", "tcp_rtomin",
                 "udp_incsumerrors",
                 "udp_indatagrams", "udp_inerrors", "udp_noports", "udp_outdatagrams",
                 "udp_rcvbuferrors",
                 "udp_sndbuferrors", "udplite_incsumerrors", "udplite_indatagrams",
                 "udplite_inerrors",
                 "udplite_noports", "udplite_outdatagrams", "udplite_rcvbuferrors",
                 "udplite_sndbuferrors"],
}

# =================================
# KAFKA SETTINGS
# =================================
KAFKA_SERVER = "{}:{}".format(os.environ.get("KAFKA_HOST", "192.168.1.175"),
                              os.environ.get("KAFKA_PORT", "9092"))
KAFKA_CLIENT_ID = 'monitoring-data-translator'
KAFKA_API_VERSION = (1, 1, 0)
KAFKA_GROUP_ID = {"openstack": "MAPE_OS_TRANSLATOR_CG",
                  "opennebula": "MAPE_OPENNEBULA_TRANSLATOR_CG",
                  "kubernetes": "MAPE_K8S_TRANSLATOR_CG",
                  "unikernels": "MAPE_UNIKERNELS_TRANSLATOR_CG",
                  "telegraf": "MAPE_VCACHE_METRICS_CG",
                  "qoe_uc1": "MAPE_VTRANSCODER_QOE_CG",
                  "generic_metrics": "MAPE_GENERIC_METRICS_CG",
                  "vtranscoder3d": "MAPE_VTRANSCODER_METRICS_CG",
                  "vtranscoder3d_spectators": "MAPE_VTRANS_SPECTATORS_CG",
                  "vce": "MAPE_VCE_METRICS_CG", }
KAFKA_MONITORING_TOPICS = {"openstack": "nfvi.*.openstack",
                           "opennebula": "nfvi.*.opennebula",
                           "kubernetes": "nfvi.*.kubernetes", # fixme before deployment
                           "unikernels": "app.unikernels.jolokia",
                           "telegraf": "app.vcache.metrics",
                           "vtranscoder3d": ['1_metrics', '2_metrics', "app.vtranscoder3d.metrics"],
                           "vtranscoder3d_spectators": "spectators.vtranscoder3d.metrics",
                           "vce": "app.vce.metrics"}
KAFKA_TRANSLATION_TOPIC = os.environ.get("KAFKA_TRANSLATION_TOPIC", "ns.instances.trans")
KAFKA_EXECUTION_TOPIC = "ns.instances.exec"
KAFKA_TIMEOUT = 10  # seconds

# =================================
# OSM SETTINGS
# =================================
OSM_IP = os.environ.get("OSM_IP", "10.100.176.66")
OSM_ADMIN_CREDENTIALS = {"username": os.environ.get("OSM_USER", "admin"),
                         "password": os.environ.get("OSM_PWD", "osmpaswword")}
OSM_COMPONENTS = {"UI": 'http://{}:80'.format(OSM_IP),
                  "NBI-API": 'https://{}:9999'.format(OSM_IP),
                  "RO-API": 'http://{}:9090'.format(OSM_IP)}

# =================================
# REDIS SETTINGS
# =================================
REDIS_HOST = os.environ.get("REDIS_IP", "10.100.176.70")
REDIS_PORT = os.environ.get("REDIS_PORT", 6379)
REDIS_NFVI_DB = 0
REDIS_PASSWORD = None
REDIS_EXPIRATION_SECONDS = os.environ.get("REDIS_EXPIRATION_SEC", 1200)  # default: 20 minutes
# REDIS_SSL = False # default
# REDIS_SSL_KEYFILE = None # default
# REDIS_CERTFILE = None # default
# REDIS_SSL_CERT_REQS = u'required' # default
# REDIS_CA_CERTS = None # default

# =================================
# INFLUXDB SETTINGS
# =================================
# See InfluxDBClient class
INFLUX_DATABASES = {
    'default': {
        'ENGINE': 'influxdb',
        'NAME': os.environ.get("INFLUXDB_DB_NAME", 'monitoring'),
        'USERNAME': os.environ.get("INFLUXDB_USER", 'root'),
        'PASSWORD': os.environ.get("INFLUXDB_PWD", 'influxdbpassword'),
        'HOST': os.environ.get("INFLUXDB_IP", "192.168.1.175"),
        'PORT': os.environ.get("INFLUXDB_PORT", 8086)
    }
}

# =================================
# GRAYLOG SETTINGS
# =================================
GRAYLOG_HOST = os.environ.get("GRAYLOG_HOST", '192.168.1.175')
GRAYLOG_PORT = int(os.environ.get("GRAYLOG_PORT", 12201))

# ==================================
# LOGGING SETTINGS
# ==================================
# See more: https://docs.python.org/3.5/library/logging.config.html

DEFAULT_HANDLER_SETTINGS = {
    'class': 'logging.handlers.RotatingFileHandler',
    'filename': "{}/logs/access.log".format(PROJECT_ROOT),
    'mode': 'w',
    'formatter': 'detailed',
    'level': 'DEBUG' if DEBUG else 'WARNING',
    'maxBytes': 4096 * 4096,
    'backupCount': 20,
}
if GRAYLOG_LOGGING:
    DEFAULT_HANDLER_SETTINGS = {
        'class': 'graypy.GELFUDPHandler',
        'formatter': 'detailed',
        'level': 'DEBUG' if DEBUG else 'WARNING',
        'host': GRAYLOG_HOST,
        'port': GRAYLOG_PORT
    }

LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'formatters': {
        'detailed': {
            'class': 'logging.Formatter',
            'format': "[%(asctime)s] - [%(name)s:%(lineno)s] - [%(levelname)s] %(message)s",
        },
        'simple': {
            'class': 'logging.Formatter',
            'format': '%(name)-15s %(levelname)-8s %(processName)-10s %(message)s'
        }
    },
    'handlers': {
        'vtranscoder3d': DEFAULT_HANDLER_SETTINGS,
        'vtranscoder3d_spectators': DEFAULT_HANDLER_SETTINGS,
        'vcache': DEFAULT_HANDLER_SETTINGS,
        'vce': DEFAULT_HANDLER_SETTINGS,
        'kubernetes': DEFAULT_HANDLER_SETTINGS,
        'opennebula': DEFAULT_HANDLER_SETTINGS,
        'openstack': DEFAULT_HANDLER_SETTINGS,
        'translator': DEFAULT_HANDLER_SETTINGS,
        'soapi': DEFAULT_HANDLER_SETTINGS,
        'errors': DEFAULT_HANDLER_SETTINGS
    },
    'loggers': {
        'vtranscoder3d': {
            'handlers': ['vtranscoder3d']
        },
        'vtranscoder3d_spectators': {
            'handlers': ['vtranscoder3d_spectators']
        },
        'vcache': {
            'handlers': ['vcache']
        },
        'vce': {
            'handlers': ['vce']
        },
        'kubernetes': {
            'handlers': ['kubernetes']
        },
        'opennebula': {
            'handlers': ['opennebula']
        },
        'openstack': {
            'handlers': ['openstack']
        },
        'translator': {
            'handlers': ['translator']
        },
        'soapi': {
            'handlers': ['soapi']
        },
        'errors': {
            'handlers': ['errors']
        }
    },
    'root': {
        'level': 'WARNING',
        'handlers': [
            'vtranscoder3d',
            'vtranscoder3d_spectators',
            'vcache',
            'vce',
            'kubernetes',
            'opennebula',
            'openstack',
            'translator',
            'soapi',
            'errors',
        ]
    },
}
